### PR TITLE
Fixing rename refactoring for record components and compact constructors

### PR DESCRIPTION
**Pull request by lahodaj**
This is an attempt to fix rename refactoring for record components. Will still need some cleanup/more testing. Does not alter the AST shapes in any way, but uses them to perform the work. Will also rename "linked" elements for the component, like the accessor and the parameter for the (compact) constructor.

@mbien - not sure if you are looking at something similar? I apologize if I stepped on something you're looking at.

### CODE REVIEW

**Review by mbien — COMMENTED** [review]
awesome! Looks like rename is working as long it is started via right click -> refactor -> rename.

the "realtime" rename mode (ctrl+R) will still produce wrong results. But this is probably linked to the semantic highlighting functionality which also doesn't associate record components with the compact constructor "params" for example.

The quickest fix for this might be to find the logic which decides if it should rename via dialog or in-place and let it open the dialog if it is a record component refactoring.

File: java/refactoring.java/src/org/netbeans/modules/refactoring/java/plugins/RenameRefactoringPlugin.java [hunk 1]
```diff
@@ -433,6 +438,27 @@ public void run(CompilationController info) throws Exception {
...
+                                            continue;
+                                        }
+                                        for (VariableElement param : c.getParameters()) {
+                                            if (param.getSimpleName().equals(el.getSimpleName())) {
+                                                //the parameter for the canonical constructor:
+                                                recordLinkedDeclarations.add(TreePathHandle.create(param, info));
+                                            }
+                                        }
+                                    }
+                                }
```
Comment (mbien) [inline java/refactoring.java/src/org/netbeans/modules/refactoring/java/plugins/RenameRefactoringPlugin.java:461]: L467 below would need the reverse case for renaming a record accessor which would have to be linked to the comonents i believe.

**Review by mbien — COMMENTED** [review]
what doesn't work yet is:
 - renaming accessors (they are not linked to the components)
 - instant-renaming the record itself will not rename the compact constructor
 
The second point is caused due to `InstntRenamePerformer#computeChangePoints` looping through `ElementFilter.constructorsIn(el.getEnclosedElements()))`, however, this list does not include the compact constructor. In fact, it isn't even in` getEnclosedElements()`- so I don't know how to get it unfortunately.

```java
public class RenameRecord {

    private record Vec2d(int x, int y) {
        public Vec2d {
            if (x < 0 || y < 0) {
                throw new IllegalArgumentException();
            }
        }
        public Vec2d(int x) {
            this(x, 0);
        }
        @Override
        public int x() {
            return x;
        }
    }

    public static void main(String[] args) {

        Vec2d vec = new Vec2d(1);
        //       rename ^^^^^  (check compact constructor during instant rename)
        //                     (this does work if the record is made public)
        
        System.out.println(vec);
        System.out.println(vec.x);
        System.out.println(vec.x());
        //              rename ^  (doesn't rename component)
        System.out.println(vec.y);
        System.out.println(vec.y());
        //              rename ^  (NPE)
    }

}
```

**Comment by mbien** [comment]
>   - instant-renaming the record itself will not rename the compact constructor
> 
> The second point is caused due to InstntRenamePerformer#computeChangePoints looping through ElementFilter.constructorsIn(el.getEnclosedElements())), however, this list does not include the compact constructor. In fact, it isn't even in getEnclosedElements()- so I don't know how to get it unfortunately.

ah I get it. The compact constructor is represented by the canonical constructor. `Utilities#getToken` can't convert it to a token though and returns null.
https://github.com/apache/netbeans/blob/36171a7e45f592c11be357fbef9a1dcf8d0e08ec/java/java.editor/src/org/netbeans/modules/java/editor/rename/InstantRenamePerformer.java#L377-L385

File: java/java.source.base/src/org/netbeans/api/java/source/ElementUtilities.java [hunk 2]
```diff
@@ -928,7 +929,143 @@ public ExecutableElement getDescriptorElement(TypeElement origin) {
...
+     *
+     * This method can be called on any {@code Element}, and will return a collection
+     * with a single entry if the provided element is not a record element.
+     *
+     * @param forElement for which the linked elements should be found
+     * @return a collection containing the provided element, plus any additional elements
+     *         that are linked to it by the Java record specification
+     * @since 2.70
+     */
+    public Collection<? extends Element> getLinkedRecordElements(Element forElement) {
```
Comment (mbien) [inline java/java.source.base/src/org/netbeans/api/java/source/ElementUtilities.java:946]: great method!

**Review by mbien — APPROVED** [review]
code looks good and it fixes all rename scenarios I could come up with - awesome!

please don't forget to squash a little before merge

**Comment by mbien** [comment]
spoke a bit too soon, using the example https://github.com/apache/netbeans/pull/7670#pullrequestreview-2243889285 and renaming a record component will rename it with instant-rename now which forgets about accessor methods. I believe last iteration used non-instant rename which works fine.

File: java/java.editor/src/org/netbeans/modules/java/editor/rename/InstantRenamePerformer.java [hunk 3]
```diff
@@ -450,14 +450,13 @@ private static boolean allowInstantRename(CompilationInfo info, Element e, Eleme
             if(isProperty) {
                 return false;
             }
-            boolean isRecordComponent = e.getEnclosingElement().getKind() == ElementKind.RECORD;
-            if (isRecordComponent) {
-                return false;
-            }
```
Comment (mbien) [inline java/java.editor/src/org/netbeans/modules/java/editor/rename/InstantRenamePerformer.java]: might be still needed after all?

Comment (lahodaj) [reply java/java.editor/src/org/netbeans/modules/java/editor/rename/InstantRenamePerformer.java]: I just placed the new check at a wrong place. Sorry for that.

**Review by mbien — APPROVED** [review]
perfect (works)

**Comment by lahodaj** [comment]
I have squashed and fixed on more issue (the rename didn't work when the rename started from the constructor's parameter, because the `TreePathHandle` didn't use the `ElementHandle` for it, which then failed the refactorings in all other files expect the one that contains the parameter). If the tests are green, and there are no objections, I would merge.

**Review by dbalek — APPROVED** [review]
Looks fine. Thanks.

**Comment by lahodaj** [comment]
I've found another problem yesterday: `ElementHandle.create` was happily creating `ElementHandle` for lambda parameters, but could not (quite obviously) resolve them back. In some cases, this then broke the refactoring. I've attempted to tweak that in this commit:
https://github.com/apache/netbeans/pull/7670/commits/56461a268907b628093ddf840cb875b329b02b79

File: java/java.source.base/apichanges.xml [hunk 4]
```diff
@@ -25,6 +25,18 @@
     <apidef name="javasource_base">Java Source API</apidef>
 </apidefs>
 <changes>
+    <change id="TreeMaker.RecordComponent">
+        <api name="javasource_base" />
+        <summary>Adding TreeMaker.RecordComponent</summary>
+        <version major="1" minor="2.70.0"/>
+        <date day="20" month="8" year="2023"/>
```
[the remark this answers is not in the document]
Comment (mbien) [reply java/java.source.base/apichanges.xml:32]: @lahodaj just wanting to make sure that 2023 was intended ;)

**Comment by mbien** [comment]
still +1 from me - great fixes!

**Pull request merged**